Added hook for when player takes item from container

cooperuser opened the pull request:
Added `net.minecraftforge.event.entity.player.ItemTakeEvent` that gets triggered from `net.minecraft.inventory.Slot#canTakeStack(EntityPlayer)` when a player clicks on an item.

I am currently making a mod called [InventoryFilter](https://github.com/cooper-anderson-mc/InventoryFilter) that allows the user to filter what items are allowed to go in their inventory. As of right now, I have item pickups following the filter, but I couldn't easily stop the player from grabbing from a container without checking the player's inventory every tick...

The new ItemTakeEvent event is cancelable, so it makes it easy to check what the item being clicked is, put it through the filter, and cancel if I need to.

Here is a basic example of how to use the new event:
```java
public class FilterEventHandler {
	@SubscribeEvent
	public void onItemTake(ItemTakeEvent event) {
		// Stop the player from being allowed to acquire TNT through an inventory.
		if (event.getItemStack().getItem().getRegistryName().toString().equals("minecraft:tnt")) {
			event.setCanceled(true);
		}
	}
}
```

cooperuser committed:
Added hook for when player takes item from container

Added `net.minecraftforge.event.entity.player.ItemTakeEvent` that gets triggered from `net.miencraft.inventory.Slot#canTakeStack(EntityPlayer)` when a player clicks on an item.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=5005) <br/>All committers have signed the CLA.

liach commented on patches/minecraft/net/minecraft/inventory/Slot.java.patch:
```diff
@@ -9,7 +9,18 @@
+     {
+-        return true;
++        ItemStack itemStack = this.func_75211_c();
++        net.minecraft.util.EnumActionResult cancelResult = net.minecraftforge.common.ForgeHooks.onItemTake(p_82869_1_, itemStack);
```
As I can see, this method is frequently overridden by mods without calling super. You might consider another place to patch so that this event remains useful.

cooperuser replied:
Is that so... That's a shame because after looking through many classes this seemed to be the easiest implementation without modifying too many files. Based on what you said about overriding, do you think it would be acceptable if I add a new method that uses the event, and calls `canTakeStack`? I would have to go and replace all the occurrences of `canTakeStack` in Minecraft's code though, but I don't think there are too many.

cooperuser replied:
Alternatively, I could just add the event code at each occurrence of `canTakeStack`, but I don't think that's the right way to go.

asiekierka replied:
I don't think you should do that either.

(a) Adding a new method means you're incompatible with all inventories which call canTakeStack on their own.
(b) Replacing each occurence won't cover mods.

I think that you should do a clean implementation and simply wait for 1.13 with it. There's no way to do this cleanly in 1.12 this late into the game.

cooperuser replied:
Hmm... That's a shame. For now do you think I should temporarily do something in my own mod as some sort of workaround for this?

loordgek commented:
you missed https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/src/main/java/net/minecraftforge/items/SlotItemHandler.java#L44

cooperuser commented:
> you missed https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/src/main/java/net/minecraftforge/items/SlotItemHandler.java#L44

I see... After looking over that method it appears that I can not use it because it doesn't have access to the `entityPlayer`. I also noticed that this class has it's own `canTakeStack ` method but I don't think I can edit that one for the same reasons as the previous one.

loordgek commented:
srry wrong one https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/src/main/java/net/minecraftforge/items/SlotItemHandler.java#L127

cooperuser commented:
I've been looking over that and reading some documentation and I am not quite sure how I actually use that function. Is it something to do with Capabilities? As far as I can see `SlotItemHandler` is not used anywhere and I'm not sure how to get it to override the base minecraft `Slot` class. Am I missing something obvious?

raoulvdberge commented:
SlotItemHandler is used by mods who use IItemHandler inventories, which are the vast majority of mods in 1.12.2.

cooperuser commented:
I see. So should the event just be triggered from both versions of `canTakeItem`? That way it will work with vanilla and modded inventories.

cooperuser committed:
Additional trigger event from `SlotItemHandler`

Added additional triggering of `ItemTakeEvent`. Many mods make use of `IItemHandler` inventories, which in turn use `SlotItemHandler` instead of `Slot`.

cooperuser committed:
Merge pull request #1 from cooper-anderson-mc/slot-item-handler

Additional trigger event from `SlotItemHandler`

cooperuser commented on src/main/java/net/minecraftforge/items/SlotItemHandler.java:
```diff
@@ -126,7 +126,8 @@ public int getItemStackLimit(@Nonnull ItemStack stack)
     public boolean canTakeStack(EntityPlayer playerIn)
     {
-        return !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        net.minecraft.util.EnumActionResult cancelResult = net.minecraftforge.common.ForgeHooks.onItemTake(playerIn, this.getStack());
```
I added the event triggering code to `SlotItemHandler#canTakeStack()`. If I understood @liach and @raoulvdberge correctly, this should do the trick.

cooperuser commented:
Are there any changes that need to be made to this?

liach reviewed (approved):
This should be fine enough.

BlayTheNinth commented:
The name is misleading, it implies an item is actually being taken when `canTakeStack` just determines if it's possible at all. Vanilla anvils call it from their GUI rendering for example. Maybe should keep the "Can" in the event name to prevent that confusion.

Also, this should probably be using `@HasResult` on the event instead of a seperate `EnumActionResult` field. EnumActionResult is generally used for item / block interactions, not for event results.

cooperuser commented:
Ah, that's a good point about the name being misleading. I guess `onCanMoveStack` would work better in that case? I just looked into how it it is used in anvils, and it appears to be used for the text color? I'm going to do a test to see if causes some unintended things to happen.

cooperuser commented:
I am also currently looking into switching to `@HasResult`

cooperuser commented:
It appears the events using `@HasResult` are in `ForgeEventFactory` instead of `ForgeHooks`. Also it appears the naming convention there would make it `onCanItemMove`? maybe `onCanStackMove`?

cooperuser committed:
Switch to `HasResult` and refactor name

cooperuser commented:
Okay I implemented the changes stated.
Here is a new version of the example mod because the name of the event changed:

```java
public class FilterEventHandler {
	@SubscribeEvent
	public void onItemTake(CanTakeItemEvent event) {
		// Stop the player from being allowed to acquire TNT through an inventory.
		if (event.getItemStack().getItem().getRegistryName().toString().equals("minecraft:tnt")) {
			event.setCanceled(true);
		}
	}
}
```

kierajreed commented:
You need to add a license header to `CanTakeItemEvent.java`... otherwise the Travis build will not pass because the license isn't there.

cooperuser committed:
Convert tabs to spaces

cooperuser commented:
Adding that now.  I can just copy it from a different file and paste it there right?

cooperuser commented:
Also I'm pretty sure I did not have a license header in the file before I renamed it... which is strange because it did not error then

liach commented:
Do `gradle licenseFormat` to fix the license.

cooperuser committed:
Add license header

cooperuser committed:
Update license header

cooperuser commented:
`gradle licenseFormat` didn't appear to do anything, or at least no changes that I could commit to fix the problem, so I added the license manually.

cooperuser commented:
I'm just wondering why I couldn't get the license error to show up on my machine

liach commented:
@cooper-anderson You need to do `gradle checkLicenses` to check licenses.

cooperuser commented:
I just tried running that command both with and without the license header and it passed both times

cooperuser commented:
![image](https://user-images.githubusercontent.com/21961037/42728695-4f1ce392-8775-11e8-9f69-efbdbbee72f3.png)

mezz commented:
The license check is passing as of your last commit, you don't need to worry about it any more.

cooperuser commented:
Sounds good. I have been reading through [Minecrell/licenser](https://github.com/Minecrell/licenser) to see if I can find anything that might be causing this to happen.

kierajreed commented:
There was a bad license in an earlier PR that was merged and later caused several other PRs' builds to fail.

cooperuser commented:
Ah, thanks for clearing that up!

cooperuser commented:
Now that all that has been sorted out, is there anything else that needs be done to this PR?

mezz commented on src/main/java/net/minecraftforge/event/ForgeEventFactory.java:
```diff
@@ -489,6 +489,13 @@ public static int onItemPickup(EntityItem entityItem, EntityPlayer player)
         return event.getResult() == Result.ALLOW ? 1 : 0;
     }
 
+    public static boolean onCanTakeItem(EntityPlayer player, ItemStack itemStack)
```
Why is this here?

mezz commented on src/main/java/net/minecraftforge/items/SlotItemHandler.java:
```diff
@@ -126,7 +126,8 @@ public int getItemStackLimit(@Nonnull ItemStack stack)
     {
-        return !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        boolean result = net.minecraftforge.event.ForgeEventFactory.onCanTakeItem(playerIn, this.getStack());
+        return result && !this.getItemHandler().extractItem(index, 1, true).isEmpty();
```
Would it be useful for the event to have a "vanilla result" it can reference?

cooperuser replied:
What do you mean exactly? Would this be something that can be accessed by an instance of the event?

Alpvax replied:
Passing the result of `!this.getItemHandler().extractItem(index, 1, true).isEmpty()` to the event, so that modders can see whether players could take the item in vanilla.

cooperuser replied:
I see. That could be useful. Is there already a convention used in other events for what this method should be called?

cooperuser replied:
Also, the `canTakeStack()` method in the vanilla `Slot` class always returns true. Should we just pass true into the event? or should we add the check manually.

Alpvax replied:
Add the check. That way if vanilla changes something down the line, we don't have to.

cooperuser replied:
Good point. Now for the name, I was thinking `getAllowed()` or something along those lines. `getVanillaResult()` would also work

cooperuser replied:
I implemented the check in `SlotItemHandler`, but I was unable to in `Slot` because `canTakeStack()` doesn't have the `ItemStack` as a parameter, and adding that would require editing lots of minecraft files and I thought it would be best to avoid that.

cooperuser committed:
Add getVanillaResult() to event

Returns true if the item is normally allowed to be moved.

mezz commented on src/main/java/net/minecraftforge/items/SlotItemHandler.java:
```diff
@@ -126,7 +126,9 @@ public int getItemStackLimit(@Nonnull ItemStack stack)
-        return !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        boolean vanillaResult = !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        boolean result = net.minecraftforge.event.ForgeEventFactory.onCanTakeItem(playerIn, this.getStack(), vanillaResult);
+        return result && vanillaResult;
```
If `vanillaResult` here is `false`, is there a point in firing the event?
I think we should carefully consider if this is the right logic, and if it is we should make it something that avoids firing when it's not required, like
`return vanillaResult && net.minecraftforge.event.ForgeEventFactory.onCanTakeItem(playerIn, this.getStack(), vanillaResult);`

cooperuser replied:
I thought about that too. But then it occurred to me that a mod maker might want to override the `false` vanilla result and force the item to be picked up. In that case should we just remove `vanillaResult` from the return statement? If not then I think making the event only trigger when it is true also works.

cooperuser replied:
Any thoughts?

mezz replied:
Let's make it so that the event result wins on this implementation. It makes it more flexible so that events can control it, but still allows mods to override it. So yeah, just remove `vanillaResult` from the return statement

cooperuser replied:
Sounds good. Besides removing `vanillaResult` from that return statement, is there anything else that we need to do?

cooperuser committed:
Remove vanillaResult from SlotItemHandler return statement

mezz commented on src/main/java/net/minecraftforge/event/ForgeEventFactory.java:
```diff
@@ -489,6 +489,13 @@ public static int onItemPickup(EntityItem entityItem, EntityPlayer player)
+    {
+        Event event = new net.minecraftforge.event.entity.player.CanTakeItemEvent(itemStack, player, vanillaResult);
+        MinecraftForge.EVENT_BUS.post(event);
+        return !event.isCanceled() && event.getResult() != Result.DENY;
```
I think this event should only use `Result` and not canceled, so that `Result.DEFAULT` means use `vanillaResult`, that way implementations like in `SlotItemHandler` are still correct.
With the way it's implemented now, `SlotItemHandler` will allow taking items by default, which changes how it works.

cooperuser replied:
So just remove the `!event.isCanceled() && ` part? or did I understand that wrong

mezz replied:
we need 3 cases:
* `Result.DENY`: return `false`
* `Result.DEFAULT`: return `vanillaResult`
* `Result.ALLOW`: return `true`

If you look at `SlotItemHandler` you can see why. Imagine nobody is listening to the event, this code is changing the method's behavior unless the `DEFAULT` action returns `vanillaResult`

Canceled is not really needed here as a way of changing the result, it's just to keep other mods from altering the result later.

cooperuser replied:
I see... So should I implement these all on one line? or have an if statement with 3 separate returns. So far I have been using one line returns, but this seems a bit complicated to read for anyone looking over the code

mezz replied:
Forge files like `ForgeEventFactory` aren't patches so they do not have line restrictions, so you can write it in a way that is more readable.

cooperuser replied:
Okay I did it using if statements for readability. Also, I put `return vanillaResult` as the default after the other cases.

cooperuser committed:
Remove isCanceled and implement different cases

mezz commented on src/main/java/net/minecraftforge/event/ForgeEventFactory.java:
```diff
@@ -489,6 +489,18 @@ public static int onItemPickup(EntityItem entityItem, EntityPlayer player)
+        MinecraftForge.EVENT_BUS.post(event);
+        if (event.getResult() == Result.DENY) {
+            return false;
+        } else if (event.getResult() == Result.ALLOW) {
```
Please follow Forge's brace style. You can use no braces for short statements like this, or put braces on a new line.

cooperuser committed:
Correct to Forge's brace style

LexManos commented on patches/minecraft/net/minecraft/inventory/Slot.java.patch:
```diff
@@ -9,7 +9,17 @@
+     {
+-        return true;
++        ItemStack itemStack = this.func_75211_c();
++        return net.minecraftforge.event.ForgeEventFactory.onCanTakeItem(p_82869_1_, itemStack, true);
```
Why is this two lines?

LexManos commented on src/main/java/net/minecraftforge/event/entity/player/CanTakeItemEvent.java:
```diff
@@ -0,0 +1,63 @@
+public class CanTakeItemEvent extends PlayerEvent
+{
+    private final ItemStack itemStack;
+    private final boolean vanillaResult;
```
So absolutly nothing else is useful in this event? Not the inventory, or the slot, or anything like that?

cooperuser committed:
Removed `this.getStack()` from it's own line

cooperuser committed:
Add `getSlot()` and `getInventory()` to CanTakeItemEvent

cooper-anderson commented:
Is there anything else that needs to be done here?

cooperuser commented:
Is there anything else that needs to be done here?

tterrag1098 commented on src/main/java/net/minecraftforge/items/SlotItemHandler.java:
```diff
@@ -126,7 +126,8 @@ public int getItemStackLimit(@Nonnull ItemStack stack)
     {
-        return !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        boolean vanillaResult = !this.getItemHandler().extractItem(index, 1, true).isEmpty();
+        return net.minecraftforge.event.ForgeEventFactory.onCanTakeItem(this.getStack(), this, this.inventory, playerIn, vanillaResult);
```
The `inventory` field is meaningless in this class. You need to provide the `IItemHandler` to the event somehow. In fact perhaps the event should *only* accept `IItemHandler` and be wrapped for vanilla cases.

tterrag1098 commented on src/main/java/net/minecraftforge/event/entity/player/CanTakeItemEvent.java:
```diff
@@ -0,0 +1,87 @@
+ *
+ * This event is fired in {@link net.minecraft.inventory.Slot#canTakeStack(EntityPlayer)} and {@link net.minecraftforge.items.SlotItemHandler#canTakeStack(EntityPlayer)} when a player clicks on an item.
+ */
+@Cancelable
```
No point having this if there is a result. You don't even use the canceled state in the event post code.
